### PR TITLE
[Profile] Refactor profile correlation.

### PR DESCRIPTION
**Pull request by ZequanWu**
Refactor some code from https://github.com/llvm/llvm-project/pull/69493.

#70712 was reverted due to linking failures. So, I removed `-profile-correlate=` flag and kept `-debug-info-correlate` in this change.

### CODE REVIEW

**Comment by ZequanWu** [comment]
Sorry for so many noise regarding this change.

**Review by ellishg — APPROVED** [review]
Sounds fine to me, but I guess I don't understand why `-profile-correlate=` doesn't work. Do you still plan to add the flag later?

**Comment by ZequanWu** [comment]
> Sounds fine to me, but I guess I don't understand why `-profile-correlate=` doesn't work. Do you still plan to add the flag later?

I haven't found a way to share information (whether of not binary correlation is enabled) between CodeGen(TargetLoweringObjectFileImpl.cpp) and Instrumentation(InstrProfiling.cpp) components. The explanation is 
here: https://discourse.llvm.org/t/rfc-add-binary-profile-correlation-to-not-load-profile-metadata-sections-into-memory-at-runtime/74565#use-temporary-section-names-6.

**Comment by dtellenbach** [comment]
@ZequanWu this seems to cause issues on macOS: https://green.lab.llvm.org/green/job/clang-stage1-RA/36184/console
```
Profile-x86_64 :: Darwin/instrprof-debug-info-correlate.c
Profile-x86_64 :: instrprof-darwin-
Profile-x86_64h :: Darwin/instrprof-debug-info-correlate.c
Profile-x86_64h :: instrprof-darwin-exports.c
```
are failing, could you please take a look or revert?

**Comment by ZequanWu** [comment]
> @ZequanWu this seems to cause issues on macOS: https://green.lab.llvm.org/green/job/clang-stage1-RA/36184/console
> 
> ```
> Profile-x86_64 :: Darwin/instrprof-debug-info-correlate.c
> Profile-x86_64 :: instrprof-darwin-
> Profile-x86_64h :: Darwin/instrprof-debug-info-correlate.c
> Profile-x86_64h :: instrprof-darwin-exports.c
> ```
> 
> are failing, could you please take a look or revert?

It should be fixed by https://github.com/llvm/llvm-project/commit/56e205a89cbb114750f2bd3f5cfbd19e209d018a.

**Comment by dtellenbach** [comment]
> It should be fixed by [56e205a](https://github.com/llvm/llvm-project/commit/56e205a89cbb114750f2bd3f5cfbd19e209d018a).

Thank you!

File: compiler-rt/lib/profile/InstrProfiling.h [hunk 1]
```diff
@@ -261,6 +261,9 @@ uint64_t __llvm_profile_get_magic(void);
 /*! \brief Get the version of the file format. */
 uint64_t __llvm_profile_get_version(void);
 
+/*! \brief If the binary is compiled with profile correlation. */
+int hasCorrelation();
```
Comment (petrhosek) [inline compiler-rt/lib/profile/InstrProfiling.h:265]: The name of this function doesn't follow the existing convention, it should be named `__llvm_profile_has_correlation`.

File: compiler-rt/lib/profile/InstrProfiling.c [hunk 2]
```diff
@@ -89,3 +89,7 @@ COMPILER_RT_VISIBILITY void __llvm_profile_reset_counters(void) {
   }
   lprofSetProfileDumped(0);
 }
+
+inline int hasCorrelation() {
```
Comment (petrhosek) [inline compiler-rt/lib/profile/InstrProfiling.c:93]: The function should have a hidden visibility (the `COMPILER_RT_VISIBILITY` macro) so it's hidden from ABI.

**Comment by ZequanWu** [comment]
> > Sounds fine to me, but I guess I don't understand why `-profile-correlate=` doesn't work. Do you still plan to add the flag later?
> 
> I haven't found a way to share information (whether of not binary correlation is enabled) between CodeGen(TargetLoweringObjectFileImpl.cpp) and Instrumentation(InstrProfiling.cpp) components. The explanation is here: https://discourse.llvm.org/t/rfc-add-binary-profile-correlation-to-not-load-profile-metadata-sections-into-memory-at-runtime/74565#use-temporary-section-names-6.

Oh, actually, we can have -profile-correlate flag just need to define it at InstrProfCorrelator.cpp (ProfileData component), which is less desired but working as discussed at https://github.com/llvm/llvm-project/pull/69656#discussion_r1372179620.

File: compiler-rt/lib/profile/InstrProfiling.h (hunk 1 above)
Comment (ZequanWu) [reply compiler-rt/lib/profile/InstrProfiling.h:265]: Renamed to `__llvm_profile_has_correlation` at https://github.com/llvm/llvm-project/commit/7fa9930847bbef4319c2d2e9c782eb5c8e6b1892

File: compiler-rt/lib/profile/InstrProfiling.c (hunk 2 above)
Comment (ZequanWu) [reply compiler-rt/lib/profile/InstrProfiling.c:93]: COMPILER_RT_VISIBILITY is added at https://github.com/llvm/llvm-project/commit/56e205a89cbb114750f2bd3f5cfbd19e209d018a.

**Comment by petrhosek** [comment]
I'm a bit concerned about the use of `hasCorrelation`. We require the runtime to check the flag and omit the data and names section if set which introduces a potential issue: since we emit the version in every TU and u se COMDAT to deduplicate them, but that means that if you link together TUs compiled with and without `-debug-info-correlate`/`-profile-correlate=` (that is having different flags), you would end up with different results depending on which section was selected by the linker. This may not be an issue if you always compile all code yourself using the same set of flags, but might be an issue when you combine libraries coming from different sources.

What I think would be a better design is to just omit the respective sections altogether when `-debug-info-correlate`/`-profile-correlate=` is enabled. Then in `llvm-profdata` we would use the correlation as a fallback mechanism, that is if the data or names are present in the profile, just use it, otherwise fallback either to debug info or unstripped file (you could even support a combination of both). That way we wouldn't even need flags like `VARIANT_MASK_DBG_CORRELATE`.

**Comment by ZequanWu** [comment]
> I'm a bit concerned about the use of hasCorrelation. We require the runtime to check the flag and omit the data and names section if set which introduces a potential issue: since we emit the version in every TU and u se COMDAT to deduplicate them, but that means that if you link together TUs compiled with and without -debug-info-correlate/-profile-correlate= (that is having different flags), you would end up with different results depending on which section was selected by the linker. This may not be an issue if you always compile all code yourself using the same set of flags, but might be an issue when you combine libraries coming from different sources.

I understand your concern. I think that's the limitation of these modes. You need to either compile all code with `-debug-info-correlate`/`-profile-correlate=` consistently or not. So, these modes are not suitable for projects which link with libraries compiled without the flags.

> What I think would be a better design is to just omit the respective sections altogether when -debug-info-correlate/-profile-correlate= is enabled.

I don't understand how this solves the problem you described above. Can you elaborate a bit more? We still need to decide if we need to omit data and names sections at runtime right?

**Comment by petrhosek** [comment]
> > What I think would be a better design is to just omit the respective sections altogether when -debug-info-correlate/-profile-correlate= is enabled.
> 
> I don't understand how this solves the problem you described above. Can you elaborate a bit more? We still need to decide if we need to omit data and names sections at runtime right?

If you omit the data and names sections from the object files, then `__llvm_profile_end_data() - __llvm_profile_begin_data()` and `__llvm_profile_end_names() - __llvm_profile_begin_names()` are both `0` and there's no need for [special casing this in the runtime](https://github.com/llvm/llvm-project/blob/24b11ba24da3e65f718391ccc85d4d22a081e893/compiler-rt/lib/profile/InstrProfilingBuffer.c#L59). When linking together objects complied with different flags, sizes of those sections would correspond only to object files compiled without correlation. For example, if linking together `a.o` compiled with debug info correlation and `b.o` compiled without, the profile would only contain data and names from `b.o`.

This approach would require slightly more complex logic in the correlator. Currently, we either read the data and names from the profile or from DWARF (depending on the flag), there's no in-between state. With the approach I'm proposing, we would need to combine data from different sources: read the data and names in the profile and combine them with the data and names in the DWARF and/or the unstripped file.

**Comment by ZequanWu** [comment]
> If you omit the data and names sections from the object files, then __llvm_profile_end_data() - __llvm_profile_begin_data() and __llvm_profile_end_names() - __llvm_profile_begin_names() are both 0 and there's no need for [special casing this in the runtime](https://github.com/llvm/llvm-project/blob/24b11ba24da3e65f718391ccc85d4d22a081e893/compiler-rt/lib/profile/InstrProfilingBuffer.c#L59).

From my experiments, I found that for debug info correlation `__llvm_profile_end_data()` - `__llvm_profile_begin_data()` and `__llvm_profile_end_names()` - `__llvm_profile_begin_names()` are both 0 at runtime because `{__start_/__stop_}{__llvm_prf_names/data}` symbols are null if there is no `__llvm_prf_names/data`. 

But for binary correlation, even though `__llvm_prf_names/data` don't have `SHF_ALLOC` flag so they are not loaded into memory, `__llvm_profile_end_data()` - `__llvm_profile_begin_data()` and `__llvm_profile_end_names()` - `__llvm_profile_begin_names()` still produces the size of name and data sections, that's why we need a bit in the global profile version variable to indicate they are empty at runtime. I feel like this is a bug in lld as __start_/__stop_ symbols for non allocated sections should be null. Even if we have a way to fix that, we still need the mode bit at llvm-profile merging step to indicate if the raw profile need debug info/binary to correlate. However, we can get ride of the runtime checking `__llvm_profile_has_correlation` to allow code compiled w/wo debug-info linked together works well.

**Comment by MaskRay** [comment]
I am nervous with new `if (hasCorrelation())` conditions as well.


If you use the same name for a non-SHF_ALLOC section and a SHF_ALLOC section, the linked output has the SHF_ALLOC flag and you cannot tell what components were non-SHF_ALLOC.
```
.section aa,"",@progbits,unique,1
.quad 0

.section aa,"a",@progbits,unique,2
.quad 0
```

Usually, the better idea is to use different section names. It may require some trial and error to see how much complexity this scheme will bring.

> I feel like this is a bug in lld as _start/_stop symbols for non allocated sections should be null. 

It isn't. This is "undefined behavior, no diagnostic required".

For
```
extern char __start___llvm_covfun __attribute__((weak));
extern char __stop___llvm_covfun __attribute__((weak));

int main() {
  printf("%p, %p\n", &__start___llvm_covfun, &__stop___llvm_covfun);
  return 0;
}
```

The program is ill-formed when `__llvm_covfun` does not have the SHF_ALLOC flag.
`__start_`/`__stop_` symbols were not designed to be used with non-SHF_ALLOC output sections.
It's invalid to reference a non-SHF_ALLOC section from SHF_ALLOC code. The regular undefined weak rule (handwavy "Unresolved weak symbols have a zero value." in the specification) does not necessarily apply.

I noticed that GNU ld defines `__start_`/`__stop_` as well and its linked a.out prints non-zero addresses, just like lld.

**Comment by ZequanWu** [comment]
Chatted with @MaskRay offline, we come to an agreement that it's not good to use relying on the bit in __llvm_profile_raw_version to decide whether or not to dump data/name sections at runtime for the reasons mentioned above. Sent a PR: #71996

> Even if we have a way to fix that, we still need the mode bit at llvm-profile merging step to indicate if the raw profile need debug info/binary to correlate. However, we can get ride of the runtime checking __llvm_profile_has_correlation to allow code compiled w/wo debug-info linked together works well.

I take back this word. When we do `llvm-profdata merge`, we use `-debug-info` to provide debug info file for correlation and the same could apply to binary correlation, a different flag. So, we know which correlation mode we are using.

**Pull request merged**